Std/Data/UnionFind/Basic.lean

```diff
@@ -165,60 +165,60 @@ def push (self : UnionFind) : UnionFind where
   rankD_lt := by simp [push_parentD, push_rankD]; exact self.rank_lt
 
 /-- Root of a union-find node -/
-def root' (self : UnionFind) (x : Fin self.size) : Fin self.size :=
+def root (self : UnionFind) (x : Fin self.size) : Fin self.size :=
   let y := (self.arr.get x).parent
   if h : y = x then
     x
   else
     have := Nat.sub_lt_sub_left (self.lt_rankMax x) (self.rank'_lt _ h)
-    self.root' ⟨y, self.parent'_lt x⟩
+    self.root ⟨y, self.parent'_lt x⟩
 termination_by _ => self.rankMax - self.rank x
 
 @[nolint unusedHavesSuffices]
-theorem parent'_root' (self : UnionFind) (x : Fin self.size) :
-    (self.arr.get (self.root' x)).parent = self.root' x := by
-  rw [root']; split <;> [assumption; skip]
+theorem parent_root (self : UnionFind) (x : Fin self.size) :
+    (self.arr.get (self.root x)).parent = self.root x := by
+  rw [root]; split <;> [assumption; skip]
   have := Nat.sub_lt_sub_left (self.lt_rankMax x) (self.rank'_lt _ ‹_›)
-  apply parent'_root'
+  apply parent_root
 termination_by _ => self.rankMax - self.rank x
 
 /-- Root of a union-find node -/
-def root (self : UnionFind) (x : Nat) : Nat :=
-  if h : x < self.size then self.root' ⟨x, h⟩ else x
+def root! (self : UnionFind) (x : Nat) : Nat :=
+  if h : x < self.size then self.root ⟨x, h⟩ else panicWith x "index out of bounds"
```

Comment:
No I want `root` too, maybe it should be called `rootD`? It is mathematically meaningful (as is `findD`), it represents the root operation when you extend it with an infinity of singleton classes / self-loops. It plays a pretty central role in a lot of the theorems here.